modules/livestore/instance.go

```diff
@@ -125,12 +135,64 @@ func newInstance(instanceID string, cfg Config, wal *wal.WAL, overrides override
 	return i, nil
 }
 
-func (i *instance) pushBytes(ts time.Time, req *tempopb.PushBytesRequest) {
+func (i *instance) backpressure(ctx context.Context) bool {
+	maxLiveTraces := i.overrides.MaxLocalTracesPerUser(i.tenantID)
+	if maxLiveTraces > 0 {
+		// Check live traces
+		i.liveTracesMtx.Lock()
+		liveTracesCount := int(i.liveTraces.Len())
+		i.liveTracesMtx.Unlock()
+
+		if liveTracesCount >= maxLiveTraces {
+			// Live traces exceeds the expected amount of data in
+			// per wal flush, so wait a bit.
+			select {
+			case <-ctx.Done():
+				return false
+			case <-time.After(1 * time.Second):
+			}
+
+			metricBackPressure.WithLabelValues(reasonWaitingForLiveTraces).Inc()
+			return true
+		}
+	}
+
+	// Check outstanding wal blocks
+	i.blocksMtx.RLock()
+	count := len(i.walBlocks)
+	i.blocksMtx.RUnlock()
+
+	if count > 1 {
+		// There are multiple outstanding WAL blocks that need completion
+		// so wait a bit.
+		select {
+		case <-ctx.Done():
+			return false
+		case <-time.After(1 * time.Second):
+		}
+
+		metricBackPressure.WithLabelValues(reasonWaitingForWAL).Inc()
+		return true
+	}
+
+	return false
+}
+
+func (i *instance) pushBytes(ctx context.Context, ts time.Time, req *tempopb.PushBytesRequest) {
 	if len(req.Traces) != len(req.Ids) {
 		level.Error(i.logger).Log("msg", "mismatched traces and ids length", "IDs", len(req.Ids), "traces", len(req.Traces))
 		return
 	}
 
+	// Wait for room in pipeline if needed
```

Comment:
Thoughts on moving the select block from inside the func to inside this for?  That way the backpressure func is only determining if backpressure should be applied and the for loop doesnt look nearly as odd.